// medusa-payment-liqpay/dist/lib/interfaces/liqpay.interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiqPayRefundRequest = exports.LiqPayInfoStatusRequest = exports.LiqPayStatusEnum = exports.LiqPayPayTypeEnum = exports.LiqPayBonusTypeEnum = exports.LiqPayActionsEnum = exports.LiqPayInfoRequestActionEnum = void 0;
var LiqPayInfoRequestActionEnum;
(function (LiqPayInfoRequestActionEnum) {
    LiqPayInfoRequestActionEnum["data"] = "data";
    LiqPayInfoRequestActionEnum["status"] = "status";
    LiqPayInfoRequestActionEnum["refund"] = "refund";
    LiqPayInfoRequestActionEnum["ticket"] = "ticket";
    LiqPayInfoRequestActionEnum["reports"] = "reports";
    LiqPayInfoRequestActionEnum["reports_compensation"] = "reports_compensation";
    LiqPayInfoRequestActionEnum["register"] = "register";
    LiqPayInfoRequestActionEnum["reports_compensation_file"] = "reports_compensation_file";
})(LiqPayInfoRequestActionEnum || (exports.LiqPayInfoRequestActionEnum = LiqPayInfoRequestActionEnum = {}));
var LiqPayActionsEnum;
(function (LiqPayActionsEnum) {
    LiqPayActionsEnum["pay"] = "pay";
    LiqPayActionsEnum["hold"] = "hold";
    LiqPayActionsEnum["subscribe"] = "subscribe";
    LiqPayActionsEnum["paydonate"] = "paydonate";
    LiqPayActionsEnum["auth"] = "auth";
    LiqPayActionsEnum["paysplit"] = "paysplit";
    LiqPayActionsEnum["regular"] = "regular";
})(LiqPayActionsEnum || (exports.LiqPayActionsEnum = LiqPayActionsEnum = {}));
var LiqPayBonusTypeEnum;
(function (LiqPayBonusTypeEnum) {
    LiqPayBonusTypeEnum["bonusplus"] = "bonusplus";
    LiqPayBonusTypeEnum["discount_club"] = "discount_club";
    LiqPayBonusTypeEnum["personal"] = "personal";
    LiqPayBonusTypeEnum["promo"] = "promo";
})(LiqPayBonusTypeEnum || (exports.LiqPayBonusTypeEnum = LiqPayBonusTypeEnum = {}));
var LiqPayPayTypeEnum;
(function (LiqPayPayTypeEnum) {
    LiqPayPayTypeEnum["card"] = "card";
    LiqPayPayTypeEnum["privat24"] = "privat24";
    LiqPayPayTypeEnum["masterpass"] = "masterpass";
    LiqPayPayTypeEnum["moment_part"] = "moment_part";
    LiqPayPayTypeEnum["cash"] = "cash";
    LiqPayPayTypeEnum["invoice"] = "invoice";
    LiqPayPayTypeEnum["qr"] = "qr";
})(LiqPayPayTypeEnum || (exports.LiqPayPayTypeEnum = LiqPayPayTypeEnum = {}));
var LiqPayStatusEnum;
(function (LiqPayStatusEnum) {
    LiqPayStatusEnum["error"] = "error";
    LiqPayStatusEnum["failure"] = "failure";
    LiqPayStatusEnum["reversed"] = "reversed";
    LiqPayStatusEnum["subscribed"] = "subscribed";
    LiqPayStatusEnum["success"] = "success";
    LiqPayStatusEnum["unsubscribed"] = "unsubscribed";
    LiqPayStatusEnum["is_3ds_verify"] = "3ds_verify";
    LiqPayStatusEnum["captcha_verify"] = "captcha_verify";
    LiqPayStatusEnum["cvv_verify"] = "cvv_verify";
    LiqPayStatusEnum["ivr_verify"] = "ivr_verify";
    LiqPayStatusEnum["otp_verify"] = "otp_verify";
    LiqPayStatusEnum["password_verify"] = "password_verify";
    LiqPayStatusEnum["phone_verify"] = "phone_verify";
    LiqPayStatusEnum["pin_verify"] = "pin_verify";
    LiqPayStatusEnum["receiver_verify"] = "receiver_verify";
    LiqPayStatusEnum["sender_verify"] = "sender_verify";
    LiqPayStatusEnum["senderapp_verify"] = "senderapp_verify";
    LiqPayStatusEnum["wait_qr"] = "wait_qr";
    LiqPayStatusEnum["p24_verify"] = "p24_verify";
    LiqPayStatusEnum["mp_verify"] = "mp_verify";
    LiqPayStatusEnum["cash_wait"] = "cash_wait";
    LiqPayStatusEnum["hold_wait"] = "hold_wait";
    LiqPayStatusEnum["invoice_wait"] = "invoice_wait";
    LiqPayStatusEnum["prepared"] = "prepared";
    LiqPayStatusEnum["processing"] = "processing";
    LiqPayStatusEnum["wait_accept"] = "wait_accept";
    LiqPayStatusEnum["wait_card"] = "wait_card";
    LiqPayStatusEnum["wait_compensation"] = "wait_compensation";
    LiqPayStatusEnum["wait_lc"] = "wait_lc";
    LiqPayStatusEnum["wait_reserve"] = "wait_reserve";
    LiqPayStatusEnum["wait_secure"] = "wait_secure";
    LiqPayStatusEnum["try_again"] = "try_again";
})(LiqPayStatusEnum || (exports.LiqPayStatusEnum = LiqPayStatusEnum = {}));
class LiqPayInfoStatusRequest {
    constructor() {
        this.version = 3;
        this.public_key = '';
        this.action = LiqPayInfoRequestActionEnum.status;
        this.order_id = '';
    }
}
exports.LiqPayInfoStatusRequest = LiqPayInfoStatusRequest;
class LiqPayRefundRequest extends LiqPayInfoStatusRequest {
    constructor() {
        super(...arguments);
        this.amount = 0.0;
    }
}
exports.LiqPayRefundRequest = LiqPayRefundRequest;
